### ISSUE TITLE
Private abstract functions are not spotted by autocxx

### ISSUE THREAD

**Issue by benesch** [google/autocxx#744]
The latest autocxx (4f34e24f1f5) fails to build protobuf-sys:

```
warning: /Users/benesch/Sites/materialize/rust-protobuf-native/target/debug/build/protobuf-sys-1d46895fccfcab95/out/autocxx-build-dir/cxx/gen0.cxx:138:5: error: field type 'google::protobuf::compiler::MultiFileErrorCollector' is an abstract class
warning:   T value;
warning:     ^
warning: /Users/benesch/Sites/materialize/rust-protobuf-native/target/debug/build/protobuf-sys-1d46895fccfcab95/out/autocxx-build-dir/cxx/gen0.cxx:1841:145: note: in instantiation of template class 'rust::MaybeUninit<google::protobuf::compiler::MultiFileErrorCollector>' requested here
warning:   ::google::protobuf::compiler::MultiFileErrorCollector *uninit = reinterpret_cast<::google::protobuf::compiler::MultiFileErrorCollector *>(new ::rust::MaybeUninit<::google::protobuf::compiler::MultiFileErrorCollector>);
warning:                                                                                                                                                 ^
warning: /Users/benesch/Sites/materialize/rust-protobuf-native/target/debug/build/protobuf-src-06001173b1f36074/out/include/google/protobuf/compiler/importer.h:202:16: note: unimplemented pure virtual method 'AddError' in 'MultiFileErrorCollector'
warning:   virtual void AddError(const std::string& filename, int line, int column,
warning:                ^
warning: /Users/benesch/Sites/materialize/rust-protobuf-native/target/debug/build/protobuf-sys-1d46895fccfcab95/out/autocxx-build-dir/cxx/gen0.cxx:138:5: error: field type 'google::protobuf::MessageFactory' is an abstract class
warning:   T value;
warning:     ^
warning: /Users/benesch/Sites/materialize/rust-protobuf-native/target/debug/build/protobuf-sys-1d46895fccfcab95/out/autocxx-build-dir/cxx/gen0.cxx:1997:107: note: in instantiation of template class 'rust::MaybeUninit<google::protobuf::MessageFactory>' requested here
warning:   ::google::protobuf::MessageFactory *uninit = reinterpret_cast<::google::protobuf::MessageFactory *>(new ::rust::MaybeUninit<::google::protobuf::MessageFactory>);
warning:                                                                                                           ^
warning: /Users/benesch/Sites/materialize/rust-protobuf-native/target/debug/build/protobuf-src-06001173b1f36074/out/include/google/protobuf/message.h:1281:26: note: unimplemented pure virtual method 'GetPrototype' in 'MessageFactory'
warning:   virtual const Message* GetPrototype(const Descriptor* type) = 0;
warning:                          ^
warning: 2 errors generated.

error: failed to run custom build command for `protobuf-sys v0.1.2+3.19.1 (/Users/benesch/Sites/materialize/rust-protobuf-native/protobuf-sys)`
```

Whereas v0.14.0 was able to produce a fair number of bindings: https://docs.rs/protobuf-sys/latest/protobuf_sys/

## Steps to Reproduce the Problem

```shell
git clone https://github.com/MaterializeInc/rust-protobuf-native
patch -p1 <<EOF
diff --git a/protobuf-sys/Cargo.toml b/protobuf-sys/Cargo.toml
index 560d18c..ebdf0a7 100644
--- a/protobuf-sys/Cargo.toml
+++ b/protobuf-sys/Cargo.toml
@@ -26,9 +26,9 @@ version = "0.1.2+3.19.1"
 edition = "2021"
 
 [dependencies]
-autocxx = "0.14.0"
+autocxx = { git = "https://github.com/google/autocxx "}
 cxx = "1.0.54"
 protobuf-src = { path = "../protobuf-src", version = "1.0.0" }
 
 [build-dependencies]
-autocxx-build = "0.14.0"
+autocxx-build = { git = "https://github.com/google/autocxx "}
EOF
cd protobuf-sys
cargo check
```

If this is too large of a reproducer, feel free to kick the can back to me, and I'll try to find the time to minimize it. I figured I'd file early, though, because @adetaylor I thought you might be able to immediately spot the problem based on the error message.

**Comment by adetaylor** [comment]
Thanks for the spot! I think I have an idea what's up. I will check it out next week.

**Comment by adetaylor** [comment]
OK, turns out I'm not so sure. I'm running a minimization run which will take about a week to complete (creduce is amazing, but realistic C++ headers have _so much stuff_ it takes a huge amount of time to strip down the test case). If you are able to come up with a minimal test case before that, great!

**Comment by adetaylor** [comment]
Minimal repro from the JSON repro case:

```cpp
namespace std {
typedef long unsigned size_t;
template <typename a, a b> struct c { static constexpr a value = b; };
using true_type = c<bool, true>;
using false_type = c<bool, false>;
typedef char string;
template <typename> class unique_ptr {};
template <typename> struct make_unique;
} // namespace std
namespace google {
namespace protobuf {
class DescriptorDatabase {
  virtual bool d() = 0;
};
namespace io {
class CodedInputStream;
class CodedOutputStream;
class ZeroCopyInputStream;
class ZeroCopyOutputStream;
} // namespace io
namespace compiler {
class Importer;
class SourceTree;
class DiskSourceTree;
} // namespace compiler
namespace util {
struct JsonParseOptions {};
struct JsonPrintOptions {};
class TimeUtil {};
} // namespace util
} // namespace protobuf
} // namespace google
namespace rust {
inline namespace cxxbridge1 {
class Str {};
} // namespace cxxbridge1
} // namespace rust
```

**Comment by benesch** [comment]
Well gosh, glad I gave up on this. I spent twenty minutes or so trying to manually minimize this last night without any luck. Thanks again for looking at this!

**Comment by adetaylor** [comment]
Yep `creduce` is a superpower. (I had to make two sets of changes to `autocxx-reduce` for it to work though... it's still a pretty flakey pipeline! So I'm glad you didn't spend 20 minutes banging your head against that instead.

I reckon that this is because the pure virtual function is private. I'm not quite sure what I changed recently which would have tripped up on that, but it should be straightforward to make a regression test case from here on, then hopefully fix it. Thanks again for the report.
